### PR TITLE
🐛 Reject nonlinear Hadamard-lifting inputs safely

### PR DESCRIPTION
**Pull request by simon1hofmann**
🤖 *AI text below* 🤖

## Summary

- Validate QCO linearity before greedy Hadamard lifting mutates the IR.
- Reject unused Pauli and CNOT outputs through pass failure.
- Verify that failure leaves the input unchanged.

Part of #2255.

## Validation

- Hadamard-lifting focused tests: 12 passed.
- Repository pre-commit hooks: passed.
- Clang-Tidy 22.1.8 on changed C++ sources: passed.

AI assistance: Codex extracted this focused change from the contract audit branch, rebased it onto current main, and ran the listed validation.

## Checklist

- [x] The pull request only contains commits that are focused and relevant to this change.
- [x] I have added appropriate tests that cover the new/changed functionality.
- [x] ~~I have updated the documentation to reflect these changes.~~ (Not applicable: no user-facing documentation change is needed.)
- [x] ~~I have added entries to the changelog for any noteworthy additions, changes, fixes, or removals.~~ (Not applicable: this focused fix is labeled skip-changelog.)
- [x] ~~I have added migration instructions to the upgrade guide (if needed).~~ (Not needed.)
- [x] The changes follow the project's style guidelines and introduce no new warnings.
- [x] The changes are fully tested and pass the CI checks.
- [x] I have reviewed my own code changes.

**If PR contains AI-assisted content:**

- [x] Any agent that created, edited, or submitted GitHub content was explicitly authorized for that scope, as required by our [AI Usage Guidelines](https://github.com/munich-quantum-toolkit/core/blob/main/docs/ai_usage.md).
- [x] Every agent-authored or agent-edited public text body begins with the visible disclosure `🤖 *AI text below* 🤖` (titles are exempt).
- [x] I have disclosed AI assistance in the PR description.
- [x] I confirm that I have personally reviewed and understood all AI-generated content, and accept full responsibility for it.

### CODE REVIEW

File: mlir/lib/Dialect/QCO/Transforms/Optimizations/HadamardLifting.cpp [hunk 1]
```diff
@@ -228,6 +229,10 @@ struct HadamardLifting final : impl::HadamardLiftingBase<HadamardLifting> {
   void runOnOperation() override {
     auto op = getOperation();
     auto* ctx = &getContext();
+    if (failed(qco::verifyLinearity(op))) {
+      signalPassFailure();
+      return;
+    }
```
Comment (burgholzer) [inline mlir/lib/Dialect/QCO/Transforms/Optimizations/HadamardLifting.cpp:235]: Hm. I am not sure we should actually be adding this in front of every pass. 
Passes can expect valid IR. They should not have to verify that it is valid.
I'd rather drop that in cases like here.
Does that make sense?

File: mlir/unittests/Dialect/QCO/Transforms/Optimizations/test_qco_hadamard_lifting.cpp [hunk 2]
```diff
@@ -81,6 +84,56 @@ class QCOHadamardLiftingTest : public testing::Test {
...
+TEST_F(QCOHadamardLiftingTest, HandlesUnusedPauliOutput) {
+  auto input = parseSourceString<ModuleOp>(R"mlir(
+    module {
+      func.func @main() {
+        %q = qco.static 0 : !qco.qubit
+        %unused = qco.x %q : !qco.qubit -> !qco.qubit
+        return
+      }
+    }
+  )mlir",
```
Comment (burgholzer) [inline mlir/unittests/Dialect/QCO/Transforms/Optimizations/test_qco_hadamard_lifting.cpp:96]: This is invalid QCO IR.
We should really enforce the invariant that passes may assume IR to be valid.
We may check in tests that this is actually true before and after passes, but the corner case checking in all of our passes can be reduced.
Even more so can the amount of tests that check these invalid IR programs.
This extends well beyond this PR.

**Review by burgholzer — CHANGES_REQUESTED** [review]
I also have my doubts with this one.

Feels to me like it would be reasonable to adjust the audit prompt. It feels like it is generating a lot of changes that we actually do not want.

**Comment by simon1hofmann** [comment]
🤖 *AI text below* 🤖

Agreed. These tests use invalid QCO IR, and Hadamard lifting may assume verified linearity. The pass-local validation and invalid-input regressions are outside the revised audit, so I am closing this PR unmerged. The broader audit criteria and disposition are recorded in #2287.
